src/internal_modules/roc_core/seqlock_impl.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2020 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/seqlock_impl.h
+//! @brief Seqlock implementation.
+
+#ifndef ROC_CORE_SEQLOCK_IMPL_H_
+#define ROC_CORE_SEQLOCK_IMPL_H_
+
+#include "roc_core/stddefs.h"
+
+namespace roc {
+namespace core {
+
+typedef uint32_t seqlock_version_t;
+
+//! Seqlock implementation.
+class SeqlockImpl {
+public:
+    //! Initialize.
+    SeqlockImpl(void* val, size_t val_size);
+
+    //! Load value version.
+    seqlock_version_t version() const;
+
+    //! Try to store value.
+    bool try_store(const void* value, seqlock_version_t& ver);
+
+    //! Store value.
+    void exclusive_store(const void* value, seqlock_version_t& ver);
+
+    //! Try to load value and version.
+    bool try_load_repeat(void* value, seqlock_version_t& ver) const;
+
+    //! Load value and version.
+    void wait_load(void* value, seqlock_version_t& ver) const;
+
+private:
+    static void
+    volatile_copy_(volatile void* dst, const volatile void* src, size_t val_size);
+
+    bool try_load_(void* value, seqlock_version_t& ver) const;
+
+    void* val_;
+    size_t val_size_;
```

Comment:
If we will store pointer and size in members, then every call (try_store, try_load, etc) will involve two additional reads from memory, and the whole operation will depend on these reads.

If, instead, we will pass `&val` and `sizeof(val)` to these methods as arguments, then chances are that no memory reads are needed: `&val` can be calculated as `this + offset`, and this is likely already in register, and `sizeof(val)` is a constant. And then on many architectures like x86_64 ARM function arguments will be passed via registers.

In our code base Seqlock is often used for small frequently used variables, like counters. E.g. seqlock is a replacement for 64-bit atomics (because 64-bit atomics are not available everywhere). So it makes sense to keep seqlock overhead low and comparable with a regular atomic.